File: ToasterLibrary/src/main/java/com/motion/toasterlibrary/productive/ReAuthInterceptor.java
package com.motion.toasterlibrary.productive;

import android.content.Context;

import com.google.gson.JsonObject;
import com.motion.toasterlibrary.productive.JsonUtils;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Body;
import retrofit2.http.POST;
import retrofit2.http.Url;

public class ReAuthInterceptor implements Interceptor {

    public interface Defination{
        @POST()
        Call<JsonObject> refresh(@Url String url, @Body JsonObject body);
    }

    private Context context;

    public ReAuthInterceptor(Context context) {
        this.context = context;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Response res = chain.proceed(chain.request());
        Request req = chain.request();

        if (res.code() == 401 && !isRestricted(req.url().toString())){
            System.out.println("ReAuth: 401 is detected");
            Retrofit core = createRetrofitClient();
        }
        return res;
    }

    private boolean isRestricted(String url){
        return false;
    }

    private  Retrofit createRetrofitClient(){
        OkHttpClient client = new OkHttpClient.Builder()
                .callTimeout(3, TimeUnit.SECONDS)
                .build();
        Retrofit core = new Retrofit.Builder()
                .baseUrl("")
                .client(client)
                .addConverterFactory(GsonConverterFactory.create(JsonUtils.gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
        return core;
    }
}

File: ToasterLibrary/src/main/java/com/motion/toasterlibrary/kotlin/ReauthInterceptor.kt
package com.motion.toasterlibrary.kotlin


import android.content.Context
import com.google.gson.JsonObject
import okhttp3.Interceptor
import okhttp3.Interceptor.Chain
import okhttp3.Response
import org.json.JSONException
import org.json.JSONObject
import retrofit2.Call
import retrofit2.http.Body
import retrofit2.http.POST
import retrofit2.http.Url
import java.io.IOException


class ReauthInterceptor
/**
 * @param context application context
 */(private val context: Context) : Interceptor {
    private interface Definition {
        @POST
        fun refresh(@Url url: String?, @Body body: JsonObject?): Call<JsonObject?>?
    }

    @Throws(IOException::class)
    override fun intercept(chain: Chain): Response {
        val res = chain.proceed(chain.request())
        val req = chain.request()
        // handle 401 case
        if (res.code() == 401) {
            try {
//                val jes = JSONObject(res.body()!!.string())
                throw IOException(res.body()!!.string())
            } catch (e: JSONException) {
                e.printStackTrace()
                throw IOException("Unexpected Error Occured ")
            }
            //            throw  new IOException(res.body().string());
        }
        return res
    }



}